### PR TITLE
updated comm tiled to have multi style ghost communication

### PR DESCRIPTION
**Pull request by Adrian-Diaz**
## Purpose

Enables comm style tiled to use multiple cutoff lengths to communicate ghosts in the borders function.

## Author(s)

Adrian Diaz (University of Florida)

## Backward Compatibility

Clear of Conflict

## Implementation Notes

Additional sendbox arrays are defined, allocated, and then defined using the multiple cutoffs available in cuttype from the Neighbor class. The changes were tested with a lead tellurium/ lead selenium interface model using both single and multi settings.

### CODE REVIEW

**Comment by rbberger** [comment]
@Adrian-Diaz I've merged in the current master which had changes to `comm_tiled.cpp`. I've tried to resolve the merge conflicts. Please review the current state and verify that it still works.

File: src/comm_tiled.cpp [hunk 1]
```diff
@@ -1874,7 +2016,12 @@ void CommTiled::allocate_swap(int n)
 
   pbc_flag = new int*[n];
   pbc = new int**[n];
-  sendbox = new double**[n];
+   //if(mode == SINGLE){
```
Comment (rbberger) [inline src/comm_tiled.cpp]: @Adrian-Diaz what's up with these comments?

Comment (Adrian-Diaz) [reply src/comm_tiled.cpp]: seems to be commented debug statements or something of the sort I forgot to delete.

**Comment by sjplimp** [comment]
@Adrian-Diaz @akohlmey I made just some cosmetic changes to CommTiled, including one for what I thought would be a memory leak if multiple runs were made.  This needs to be merged with current master.  Adrian if you can then check this works as you expect, then it should be ready to merge.  Also Adrian, do you have a non-CAC test problem for running a triclinic geometry with multi-neigh and multi-comm?  If so, it would be good to add it the examples dir, since it exercises several options together.

**Comment by akohlmey** [comment]
@sjplimp @Adrian-Diaz I just updated the pull request to current master. Also added tests to the unittest facility for pair styles, that tests the colloid pair style with regular neighbor lists and communication and then multi-cutoff neighbor lists with multi-cutoff model communication and both brick and tiled style. All equivalent reference forces and stresses across the 3 tests files agree to machine precision. Fixed a couple of bugs causing segmentation faults in the process.

**Comment by Adrian-Diaz** [comment]
@akohlmey  Thanks for the patch and tests. It looks like the only thing missing is making this version of comm_tiled.cpp compatible with the recent changes enabling triclinic boxes. I'm taking a look at that now.

**Comment by akohlmey** [comment]
> @akohlmey Thanks for the patch and tests. It looks like the only thing missing is making this version of comm_tiled.cpp compatible with the recent changes enabling triclinic boxes. I'm taking a look at that now.

For testing, you could try making copies of the existing YAML files and add a `change_box all triclinic` to the post_command section. That should give the same results when the tilt factors are 0.0 (but will have more noise, so probably the epsilon setting in the test files may need to be increased a little bit.

**Comment by akohlmey** [comment]
@Adrian-Diaz @sjplimp removed the new merge conflict and added a test for the now enabled multi-cutoff with triclinic.

**Comment by Adrian-Diaz** [comment]
It looks good to me.

**Comment by sjplimp** [comment]
@Adrian-Diaz @akohlmey Sanity check.  Adrian has made 3 changes to CommTiled, 2 of which were in recently merged PRs.  (1) Support for multi neigh lists (this one).  (2) Bug fix for self comm issue.  (3) Support for triclinic.  Are you saying that this CommTiled file should now include all 3 changes?  And Adrian, have you run any of your tests, like the one for self-comm, to verify that this new merged version still passes them?

**Comment by akohlmey** [comment]
@sjplimp yes, this branch should now have all changes. It passes (serial) tests with an input deck using the colloid pair style with comm_style tiled, using multi-cutoff communication in orthogonal and triclinic boxes. I also have successfully run a similar input script in parallel. You can compare the .yaml files with "colloid" in the folder unittest/force-styles/tests/ and the listed stresses and forces should all be the same modulo numerical noise.

**Comment by Adrian-Diaz** [comment]
@sjplimp Yea all the necessary changes for those PRs are included in this version now. Most of them came in with the merge and I double checked setup, borders, and exchange for the changes and added what was missing for triclinic.

**Comment by sjplimp** [comment]
@Adrian-Diaz @akohlmey My latest push does 2 things: (1) add back a couple changes that I think were lost in the merge.  Namely (a) set nswap later in init(), see comment, (b) use cutghost[] instead of cut in SINGLE mode in borders() - I think this was there to support triclinic.  And (2) change the logic in borders() to use Waitany() instead of Waitall() and put the self comm back before processing of receives.  This is enabled by a change in AtomVec::unpack_borders() to insure the peratom data is realloced as long as needed to support the current unpack.  Adrian suggested something like this previously, instead of his bugfix for this in CommTiled.  And I think doing it this way is more efficient than Waitall() and delaying the self copies.  If you both agree with these changes (and all the tests pass), then this should be ready to merge.

**Comment by akohlmey** [comment]
@sjplimp had to remove some tabs (you probably should set the variable `indent-tabs-mode` in your emacs configuration to `nil`)
and trailing whitespace. After compilation, the code passes the bundled (serial) unit tests without further adjustments.

**Comment by Adrian-Diaz** [comment]
@sjplimp The changes look good to me and I can't think of anything wrong with them. Sorry I missed those cut variables :).

**Pull request merged**